README
CoverageTestRunner
==================

The Python standard library module `unittest` aids in implementing and running
unit tests. The `coverage.py` module (not in the standard library) measures
which parts of a program are executed. A proper unit test suite will
execute all parts of a program. `CoverageTestRunner` uses `coverage.py` to
run test suites implemented with `unittest`, and fails the tests if they
do not test the entire program.

`CoverageTestRunner` makes some assumptions, the most important is
that the test suite is written so that each module corresponds to a test
module, and that test module should execute everything in the module.
If there is a code module `foo.py`, then it must be fully tested by
`foo_tests.py`.

Ideally, we would test each class and its corresponding test suite, but
unfortunately `coverage.py` only allows module granularity. It also only
measures at the statement level. This is, however, typically better
than can be achieved without any measurement tools.


Example
-------

Assume a Python module in `foo.py`, and its test suite in `foo_tests.py`
in the same directory.
Then running the test suite using `CoverageTestRunner` can be very simple:

    python -m CoverageTestRunner

`CoverageTestRunner` scans the current 
directory and its subdirectories for test modules with corresponding
code modules, based on filenames, and then executes them pairwise.

If you have a more complicated setup, you can pair the code and test
modules manually, using a wrapper script such as follows:

    from CoverageTestRunner import CoverageTestRunner
    
    r = CoverageTestRunner()
    r.add_pair("code/foo.py", "tests/foo.py")
    r.run()


Output
------

`CoverageTestRunner` writes its output to the standard output, and for
a successful run it looks like this:

    Running test 2/2: testTrue (foo_tests.FooTests) 
    
    OK
    Time: 0.0 s

An unsuccessful run might look like this (shortened for space reasons):

    Running test 226/226: test (configTests.WriteDefaultConfigTests)        
    
    FAILED
    
    ERROR: testEmptyFile (ioTests.FileContentsTests)
    Traceback (most recent call last):
      File "unittests/ioTests.py", line 228, in testEmptyFile
        id = obnam.io.create_file_contents_object(self.context, filename)
      File "/home/liw/Braawi/obnam/new-storage/obnam/io.py", line 240, in create_file_contents_object
        stdout=subprocess.PIPE)
      File "subprocess.py", line 593, in __init__
        errread, errwrite)
      File "subprocess.py", line 1135, in _execute_child
        raise child_exception
    OSError: [Errno 2] No such file or directory

    Statements missed by per-module tests:
      Module                                                 Missed statements
      /home/liw/Braawi/obnam/new-storage/obnam/cfgfile.py    70, 303
      /home/liw/Braawi/obnam/new-storage/obnam/cmp.py        155
      /home/liw/Braawi/obnam/new-storage/obnam/config.py     234-237
      /home/liw/Braawi/obnam/new-storage/obnam/filelist.py   88
      /home/liw/Braawi/obnam/new-storage/obnam/format.py     74, 97
    
    1 failures, 3 errors
    Time: 5.4 s

In this case, after the line with `FAILED` comes a list of all the
failed tests, with some indication of why they failed, complete with
Python stack traces. Additionally, if some parts of the code were not
executed during the tests, a list of files and statement numbers is
included in the output.


Excluding some statements from the coverage analysis
----------------------------------------------------

Use `# pragma: no cover` to exclude some statements from being included
in the coverage analysis. The line with that comment is excluded. If
the line is an `if`, `while`, `def`, or similar, with a body, the body
is also excluded. For example:

    if always_true_during_testing:
        print foo
    else: # pragma: no cover
        print bar
        
The last two lines of codes are excluded from coverage analysis even
if the test suite does not execute them.

You should only use this after thorough consideration.


Excluding whole modules
-----------------------

Sometimes there are code modules that should not have test modules.
For example, `setup.py`. You can use the `--ignore-missing` option
to ignore all such problem, or `--ignore-missing-from=FILE` to
list excluded modules explicitly.

By default, the `setup.py` module is automatically excluded. However,
if you list any modules explicitly, you should also list `setup.py`.


Your tests are too slow
-----------------------

`CoverageTestRunner` keeps track of how long each test takes. If the total
run time of the tests is more than ten seconds, even if all tests pass,
it then prints out a list of the slowest tests. This is useful for projects
whose test suites should run quickly so that developers don't mind running
them all the time.

This is obviously a bad thing for projects with large test suites. If it
bothers you, please ask me, and I'll provide a way to disable it, or to
set the time limit.


Contacts
--------

Python's `unittest` is part of the standard library, and was (originally?)
written by Steve Purcell. `coverage.py` was originally written by Gareth
Rees and is now maintained by Net Batchelder. Its home page is at

    http://www.nedbatchelder.com/code/modules/coverage.html

CoverageTestRunner was written by Lars Wirzenius (`liw@liw.fi`). 
The home page is at

    http://liw.fi/coverage-test-runner/

I welcome any feedback you may have.


License
--------

Copyright 2007-2013  Lars Wirzenius <liw@liw.fi>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
 
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
 
You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

=*= License: GPL-3+ =*=



NEWS
NEWS about CoverageTestRunner
=============================

Version 1.13.1, released 2017-07-05
----------------------------------

* Fix project name in setup.py to match the Debian source package
  name.

Version 1.13, released 2017-07-05
---------------------------------

* Added Python3 support to the Debian packaging.

Version 1.11, released 2015-10-25
--------------------------------

* Debian packaging changes.

Version 1.10, released 2013-04-17
--------------------------------

* New release to trigger builds for all relevant targets.
* Copyright license summary markup added.

Version 1.9, released 2012-09-23
--------------------------------

* Home page address updated.

Version 1.8, released 2012-01-19
--------------------------------

* Progress reporting is now disabled when output does not go to a terminal.
  This will make build logs prettier.
* `logging` output is now disabled when unit tests are run. This means that
  individual projects no longer need to hide the output themselves.

Version 1.7.1, released 2011-09-04
--------------------------------

* Bug fix: normalize all relevant paths, when checking for ignored modules.

Version 1.7, released 2011-09-04
--------------------------------

* When running "python -m CoverageTestRunner", user may now specify a list
  of directories to scan for tests, rather than always scanning the current
  working directory.
* Pathnames in the file specified with `--ignore-missing-from=FILE` are
  now normalized, so that the user does not need to specify `./setup.py`
  and can say `setup.py` instead.

Version 1.6, released 2011-06-03
--------------------------------

* License change to GPL version 3 or later.
* Debian packaging updated for upload to Debian.
* `README` markup fixes and other minor updates.
* `setup.py` fixed up so it is suitable for PyPI use.

Version 1.5, released 2011-02-18
--------------------------------

* Report number of excluded modules.
* Ignore setup.py by default.

Version 1.4.1, released 2010-06-07
--------------------------------

* Allow use of profiling (via cProfile), by setting 
  COVERAGE_TEST_RUNNER_PROFILE environment variable. This will let you
  find out why the test suite is slow.


Version 1.3, released 2009-11-18
--------------------------------

* Threshold for when slow test timings are reported is now user settable
  via the COVERAGE_TEST_RUNNER environment variable.


Version 1.2, released 2009-11-18
--------------------------------

* Filenames are no longer prefixed by path to current directory.

* Number of excluded statements is reported.

* Modules without test modules are reported.

* .deb packaging included in upstream source.

* Option --ignore-missing-from added.


Version 1.1, released 2008-05-18
--------------------------------

Added support for '#pragma: no cover'.


subdir/foo_tests.py
import unittest, foo

class FooTests(unittest.TestCase):

    def testTrue(self):
        f = foo.Foo()
        self.failUnlessEqual(f.foo(True), True)
    
    def testFalse(self):
        f = foo.Foo()
        self.failUnlessEqual(f.foo(False), False)


CoverageTestRunner.py
# Copyright (C) 2007-2013  Lars Wirzenius <liw@iki.fi>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# =*= License: GPL-3+ =*=

import coverage
import unittest
import optparse
import os
import imp
import sys
import time
import logging


__version__ = '1.11'


class AllowNothing(logging.Filter):

    '''A logging library filter that disables everything.'''

    def filter(self, record):
        return False


class CoverageTestResult(unittest.TestResult):

    def __init__(self, output, total):
        unittest.TestResult.__init__(self)
        self.output = output
        self.total = total
        self.lastmsg = ""
        self.coverage_missed = []
        self.coverage_excluded = []
        self.timings = []
        self.missing_test_modules = []

    def addCoverageMissed(self, filename, statements, missed_statements,
                          missed_description):
        self.coverage_missed.append((filename, statements, missed_statements,
                                     missed_description))

    def addCoverageExcluded(self, statements):
        self.coverage_excluded += statements

    def addMissingTestModule(self, modulepath):
        self.missing_test_modules.append(modulepath)

    def wasSuccessful(self, ignore_coverage=False, ignore_missing=False):
        return (unittest.TestResult.wasSuccessful(self) and
                (ignore_coverage or not self.coverage_missed) and
                (ignore_missing or not self.missing_test_modules))

    def _ttywrite(self, string):
        if self.output.isatty():
            self.output.write(string)
            self.output.flush()

    def clearmsg(self):
        self._ttywrite("\b \b" * len(self.lastmsg))
        self.lastmsg = ""

    def write(self, test):
        self.clearmsg()
        self.lastmsg = "Running test %d/%d: %s" % (self.testsRun,
                                                   self.total,
                                                   str(test)[:50])
        self._ttywrite(self.lastmsg)

    def startTest(self, test):
        unittest.TestResult.startTest(self, test)
        self.write(test)
        self.start_time = time.time()

    def stopTest(self, test):
        end_time = time.time()
        unittest.TestResult.stopTest(self, test)
        self.timings.append((end_time - self.start_time, test))


class CoverageTestRunner:

    """A test runner class that insists modules' tests cover them fully."""

    def __init__(self):
        self._dirname = None
        self._module_pairs = []
        self._missing_test_modules = []
        self._excluded_modules = []
        self.allow_nothing = AllowNothing()

    def add_pair(self, module_pathname, test_module_pathname):
        """Add a module and its test module to list of tests."""
        self._module_pairs.append((module_pathname, test_module_pathname))

    def add_missing(self, module_pathname):
        self._missing_test_modules.append(module_pathname)

    def add_excluded_module(self, module_pathname):
        self._excluded_modules.append(module_pathname)

    def find_pairs(self, dirname, ignored_modules):
        """Find all module/test module pairs in directory tree.

        This method relies on a naming convention: it scans a directory
        tree and assumes that for any file foo.py, if there exists a
        file foo_tests.py or fooTests.py, they form a pair.

        """

        suffixes = ["_tests.py", "Tests.py"]

        self._dirname = os.path.abspath(dirname)
        if not self._dirname.endswith(os.sep):
            self._dirname += os.sep

        for dirname, dirnames, filenames in os.walk(dirname):
            filenames = [x for x in filenames if x.endswith(".py")]

            tests = []
            for suffix in suffixes:
                tests += [(x, x[:-len(suffix)] + ".py")
                          for x in filenames if x.endswith(suffix)]

            nontests = []
            nontests = [x for x in filenames
                        if x not in [a for a, b in tests]]

            for filename, module in tests:
                if module in nontests:
                    nontests.remove(module)
                    module = os.path.join(dirname, module)
                    filename = os.path.join(dirname, filename)
                    self.add_pair(module, filename)

            for filename in nontests:
                filename = os.path.normpath(os.path.join(dirname, filename))
                if filename not in ignored_modules:
                    self.add_missing(filename)
                else:
                    self.add_excluded_module(filename)

    def _load_module_from_pathname(self, pathname):
        for tuple in imp.get_suffixes():
            suffix, mode, type = tuple
            if pathname.endswith(suffix):
                name = os.path.basename(pathname[:-len(suffix)])
                f = open(pathname, mode)
                return imp.load_module(name, f, pathname, tuple)
        raise Exception("Unknown module: %s" % pathname)

    def _load_pairs(self):
        module_pairs = []
        loader = unittest.defaultTestLoader
        for pathname, test_pathname in self._module_pairs:
            module = self._load_module_from_pathname(pathname)
            test_module = self._load_module_from_pathname(test_pathname)
            suite = loader.loadTestsFromModule(test_module)
            module_pairs.append((module, test_module, suite))
        return module_pairs

    def printErrorList(self, flavor, errors):
        for test, error in errors:
            print("%s: %s" % (flavor, str(test)))
            print(str(error))

    def enable_logging(self):
        logger = logging.getLogger()
        logger.removeFilter(self.allow_nothing)

    def disable_logging(self):
        logger = logging.getLogger()
        logger.addFilter(self.allow_nothing)

    def run(self):
        start_time = time.time()

        module_pairs = self._load_pairs()
        total_tests = sum(suite.countTestCases()
                          for x, y, suite in module_pairs)
        result = CoverageTestResult(sys.stdout, total_tests)

        for path in self._missing_test_modules:
            result.addMissingTestModule(path)

        # coverage.Coverage is a coverage.py 4.x feature.
        if hasattr(coverage, 'Coverage'):
            _coverage = coverage.Coverage()
        else:
            _coverage = coverage

        for module, test_module, suite in module_pairs:
            _coverage.erase()
            _coverage.exclude(r"#\s*pragma: no cover")
            _coverage.start()
            sys.path.insert(0, os.path.dirname(module.__file__))
            imp.reload(module)
            del sys.path[0]
            self.disable_logging()
            suite.run(result)
            self.enable_logging()
            _coverage.stop()
            filename, stmts, excluded, missed, missed_desc = \
                _coverage.analysis2(module)
            if self._dirname and filename.startswith(self._dirname):
                filename = filename[len(self._dirname):]
            if missed:
                result.addCoverageMissed(filename, stmts, missed, missed_desc)
            result.addCoverageExcluded(excluded)

        end_time = time.time()

        sys.stdout.write("\n\n")

        if result.wasSuccessful():
            print("OK")
        else:
            print("FAILED")
            print()
            if result.errors:
                self.printErrorList("ERROR", result.errors)
            if result.failures:
                self.printErrorList("FAILURE", result.failures)
            if result.coverage_missed:
                print ("Statements missed by per-module tests:")
                width = max(len(x[0]) for x in result.coverage_missed)
                fmt = "  %-*s   %s"
                print(fmt % (width, "Module", "Missed statements"))
                for filename, _, _, desc in sorted(result.coverage_missed):
                    print(fmt % (width, filename, desc))
                print()
            if result.missing_test_modules:
                print("Modules missing test modules:")
                for pathname in result.missing_test_modules:
                    print("  %s" % pathname)
                print()

            print("%d failures, %d errors" % (
                len(result.failures), len(result.errors)))

        if result.coverage_excluded:
            print(len(result.coverage_excluded), "excluded statements")
        if self._excluded_modules:
            print(len(self._excluded_modules), "excluded modules")
        if result.missing_test_modules:
            print(len(result.missing_test_modules), "missing test modules")

        maxtime = int(os.environ.get('COVERAGE_TEST_RUNNER_MAX_TIME', '10'))
        if end_time - start_time > maxtime:
            print()
            print("Slowest tests:")
            for secs, test in sorted(result.timings)[-10:]:
                print("  %5.1f s %s" % (secs, str(test)[:70]))

        print("Time: %.1f s" % (end_time - start_time))

        return result


def run():
    """Use CoverageTestRunner on the desired directory."""

    parser = optparse.OptionParser()
    parser.add_option("--ignore-coverage", action="store_true",
                      help="Don't fail tests even if coverage is "
                           "incomplete.")
    parser.add_option("--ignore-missing", action="store_true",
                      help="Don't fail even if some modules have no test "
                           "module.")
    parser.add_option("--ignore-missing-from", metavar="FILE",
                      help="Ignore missing test modules for modules listed "
                           "in FILE.")

    opts, dirnames = parser.parse_args()
    if not dirnames:
        dirnames = ['.']

    if opts.ignore_missing_from:
        lines = open(opts.ignore_missing_from).readlines()
        lines = [x.strip() for x in lines]
        lines = [x for x in lines if x and not x.startswith('#')]
        lines = [os.path.normpath(x) for x in lines]
        ignored_modules = lines
    else:
        ignored_modules = ['setup.py']

    runner = CoverageTestRunner()
    for dirname in dirnames:
        runner.find_pairs(dirname, ignored_modules)
    result = runner.run()
    if not result.wasSuccessful(ignore_coverage=opts.ignore_coverage,
                                ignore_missing=opts.ignore_missing):
        sys.exit(1)


if __name__ == "__main__":
    profname = os.environ.get('COVERAGE_TEST_RUNNER_PROFILE', None)
    if profname is None:
        run()
    else:
        import cProfile
        cProfile.run('run()', profname)
